Desi/review simulator

desi-ivanova opened the pull request.

desi-ivanova committed:
remove the sign filter

desi-ivanova committed:
fix the proportions passed

jogonba2 commented:
looks good to me. @desi-ivanova if you consider it is correct enough now, just merge it and we can recompute the plots :)

desi-ivanova commented:
there's still some discrepancy for small N and small delta. I want to check a few other things

desi-ivanova commented:
there's a weird bias that's very clearly visible in small delta, small sample sizes
![image](https://github.com/user-attachments/assets/63376597-dd31-46b5-a46e-5ccd0730ca20)
Looking into fixing that

Update 1: 

For tracking purposes: I think it might be because both tests complain occasionally about bad data (division by 0)

- std is 0 in the z-test
```
/Users/zizi/miniconda3/lib/python3.12/site-packages/statsmodels/stats/weightstats.py:792: RuntimeWarning: invalid value encountered in scalar divide
  zstat = value / std
```

- same for Mcnmer
```
/Users/zizi/miniconda3/lib/python3.12/site-packages/statsmodels/stats/contingency_tables.py:1348: RuntimeWarning: divide by zero encountered in scalar divide
  statistic = (np.abs(n1 - n2) - corr)**2 / (1. * (n1 + n2))
```

Update 2:
The above shouldn't cause problems in the first case since np.nan < alpha returns False (i.e the null is, correctly, not rejected).

Haven't looked into the second case yet.

Update 3:
Changing seed improve things a bit (we were generating data with the same seed)
I think the small discrepancy is probably due to the discreteness of the problem, which is more obvious at small sample sizes

desi-ivanova commented:
Update 4:
Ok, I'm finally happy with it.
![image](https://github.com/user-attachments/assets/c20a671b-3020-4dfd-879f-a555c0ea16d1)


The discrepancy was due to the data generating process we've assumed. Actually just summing across respective rows and cols doesn't give us the independent samples, we have to sample the two binomials independently. Otherwise we end up with non-representative samples due to the correlation that's introduced hence the gap that was still remaining. 

I'll clean up and push the changes.

desi-ivanova committed:
fix DGP and refactor things a bit

desi-ivanova committed:
update test_power post refactor

desi-ivanova committed:
black test_speed. Test fails to run though

desi-ivanova committed:
fix test_stat_test post refactor

desi-ivanova committed:
cosmetic

desi-ivanova committed:
rm main from test

desi-ivanova committed:
handle the invalid tests properly.

For the ztest, the result is nan and hence nan < alpha returns always False, i.e. correctly rejects the null
Hoewever McNemar test returns inf when b+c is 0 and a pvalue of 0, i.e. always incorrectly rejects the null.

desi-ivanova committed:
update plot

desi-ivanova commented:
![image](https://github.com/user-attachments/assets/6e50c72f-797b-4aff-9e60-88846d2a3875)

Fixed the non-monotonically increasing power with delta issue for the McNemar test. Indeed the problem was dividing by 0 in the denom (noted in Update 1 above), which is handled badly in the implementation of that test (unlike the z test which is reasonable).

@anbasile @jogonba2 @asarvazyan  Please have a look. 

I was hoping that you would have  resolved all these issues by the start of the week so I could spend today and tomorrow writing. Given I spent all of my free time today debugging, this means unfortunately we won't make the deadline. It's totally fine, we can submit to the next one (COLM maybe?)

For our meeting on Tuesday, let's please have everything tidied up, plots included in the overleaf etc.

anbasile commented:
Ah! Thanks for the amazing work Desi.

Yes, we missed the deadline but I think us 3 alone would not have even seen the problem and we would have hand-waved something. Thanks a lot!

The PR with code for the plots is in already. We'll merge this one first, then that one.

COLM is a nice option. We have some ideas for a higher level API. Let's talk on Tuesday.

desi-ivanova commented:
> looks good to me. @desi-ivanova if you consider it is correct enough now, just merge it and we can recompute the plots :)

btw if you prepare the scripts (test them on coarse grid, small num of iterations etc), I'm happy to run them on my machine -- let me know

anbasile commented on power/types.py:
```diff
@@ -16,10 +16,32 @@ class DGPParameters:
     dataset_size: int
 
+    @property
+    def success_probs(self) -> np.ndarray:
```
Not 100% sure about this here. Might be too generic. Let's merge as is and review later.

anbasile reviewed (approved):
LGTM. Let's merge.
